File: gofiber/domain/services/face_service.go
package services

import (
	"context"
	"errors"

	"github.com/google/uuid"

	"gofiber-template/domain/models"
)

// Custom errors for face service
var (
	ErrNoFacesDetected   = errors.New("no faces detected in the uploaded image")
	ErrFaceNotFound      = errors.New("face not found")
	ErrInvalidFaceIndex  = errors.New("invalid face index")
)

// FaceSearchResult represents a face search result
type FaceSearchResult struct {
	Face       models.Face
	Photo      models.Photo
	Similarity float64
}

// DetectedFace represents a face detected in an uploaded image
type DetectedFace struct {
	Index      int       `json:"index"`
	BboxX      float64   `json:"bbox_x"`
	BboxY      float64   `json:"bbox_y"`
	BboxWidth  float64   `json:"bbox_width"`
	BboxHeight float64   `json:"bbox_height"`
	Confidence float64   `json:"confidence"`
	Embedding  []float32 `json:"-"` // Hidden from JSON, used internally
}

// FaceService handles face-related operations
type FaceService interface {
	// Detect faces in an uploaded image (returns all faces with bounding boxes)
	DetectFaces(ctx context.Context, imageData []byte, mimeType string) ([]DetectedFace, error)

	// Search by uploading a photo with face index selection
	SearchByImageWithIndex(ctx context.Context, userID uuid.UUID, imageData []byte, mimeType string, faceIndex int, limit int, threshold float64) ([]FaceSearchResult, error)

	// Search by uploading a photo (uses first face - legacy)
	SearchByImage(ctx context.Context, userID uuid.UUID, imageData []byte, mimeType string, limit int, threshold float64) ([]FaceSearchResult, error)

	// Search by existing face ID
	SearchByFaceID(ctx context.Context, userID uuid.UUID, faceID uuid.UUID, limit int, threshold float64) ([]FaceSearchResult, error)

	// Get faces for a photo
	GetFacesByPhoto(ctx context.Context, userID uuid.UUID, photoID uuid.UUID) ([]models.Face, error)

	// Get faces for a person
	GetFacesByPerson(ctx context.Context, userID uuid.UUID, personID uuid.UUID) ([]models.Face, error)

	// Get all faces with pagination
	GetFaces(ctx context.Context, userID uuid.UUID, page, limit int) ([]models.Face, int64, error)

	// Assign face to a person
	AssignFaceToPerson(ctx context.Context, userID uuid.UUID, faceID uuid.UUID, personID uuid.UUID) error

	// Remove face from person
	RemoveFaceFromPerson(ctx context.Context, userID uuid.UUID, faceID uuid.UUID) error

	// Get face count
	GetFaceCount(ctx context.Context, userID uuid.UUID) (int64, error)

	// Get processing stats
	GetProcessingStats(ctx context.Context, userID uuid.UUID) (*FaceProcessingStats, error)

	// Retry failed photos
	RetryFailedPhotos(ctx context.Context, userID uuid.UUID, folderID *uuid.UUID) (int64, error)
}

// FaceProcessingStats contains face processing statistics
type FaceProcessingStats struct {
	TotalPhotos     int64 `json:"total_photos"`
	ProcessedPhotos int64 `json:"processed_photos"`
	PendingPhotos   int64 `json:"pending_photos"`
	FailedPhotos    int64 `json:"failed_photos"`
	TotalFaces      int64 `json:"total_faces"`
}

File: vite/src/services/face/face.service.ts
/**
 * Face Service - API calls for face detection and search
 */
import axios from 'axios'
import { apiClient, FACE_API, type ApiResponse } from '@/shared/lib/api'
import type {
  FaceSearchData,
  FaceProcessingStats,
  FaceListData,
  Face,
  DetectedFacesData,
} from '@/shared/types'

export const faceService = {
  /**
   * Detect all faces in an uploaded image
   * Returns bounding boxes for face selection
   */
  detectFaces: async (
    imageFile: File
  ): Promise<ApiResponse<DetectedFacesData>> => {
    const formData = new FormData()
    formData.append('image', imageFile)

    try {
      const response = await apiClient.post<ApiResponse<DetectedFacesData>>(
        FACE_API.DETECT,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      )
      return response.data
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.data) {
        return error.response.data as ApiResponse<DetectedFacesData>
      }
      throw error
    }
  },

  /**
   * Search faces by uploading an image
   * @param faceIndex - Index of the face to search (default: 0)
   */
  searchByImage: async (
    imageFile: File,
    limit = 20,
    threshold = 0.6,
    faceIndex = 0
  ): Promise<ApiResponse<FaceSearchData>> => {
    const formData = new FormData()
    formData.append('image', imageFile)

    const params = new URLSearchParams({
      limit: limit.toString(),
      threshold: threshold.toString(),
      face_index: faceIndex.toString(),
    })

    try {
      const response = await apiClient.post<ApiResponse<FaceSearchData>>(
        `${FACE_API.SEARCH_BY_IMAGE}?${params}`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      )
      return response.data
    } catch (error) {
      // Handle axios errors - extract message from response
      if (axios.isAxiosError(error) && error.response?.data) {
        return error.response.data as ApiResponse<FaceSearchData>
      }
      throw error
    }
  },

  /**
   * Search faces by existing face ID
   */
  searchByFaceId: async (
    faceId: string,
    limit = 20,
    threshold = 0.6
  ): Promise<ApiResponse<FaceSearchData>> => {
    const response = await apiClient.post<ApiResponse<FaceSearchData>>(
      FACE_API.SEARCH_BY_FACE,
      {
        face_id: faceId,
        limit,
        threshold,
      }
    )

    return response.data
  },

  /**
   * Get face processing statistics
   */
  getStats: async (): Promise<ApiResponse<FaceProcessingStats>> => {
    const response = await apiClient.get<ApiResponse<FaceProcessingStats>>(
      FACE_API.STATS
    )

    return response.data
  },

  /**
   * Get paginated list of faces
   */
  getFaces: async (
    page = 1,
    limit = 50
  ): Promise<ApiResponse<FaceListData>> => {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
    })

    const response = await apiClient.get<ApiResponse<FaceListData>>(
      `${FACE_API.LIST}?${params}`
    )

    return response.data
  },

  /**
   * Get faces for a specific photo
   */
  getFacesByPhoto: async (
    photoId: string
  ): Promise<ApiResponse<{ faces: Face[]; count: number }>> => {
    const response = await apiClient.get<ApiResponse<{ faces: Face[]; count: number }>>(
      FACE_API.BY_PHOTO(photoId)
    )

    return response.data
  },

  /**
   * Retry failed face processing
   * Resets failed photos to pending for reprocessing
   */
  retryFailed: async (
    folderId?: string
  ): Promise<ApiResponse<{ reset_count: number }>> => {
    const params = folderId ? `?folder_id=${folderId}` : ''
    const response = await apiClient.post<ApiResponse<{ reset_count: number }>>(
      `${FACE_API.RETRY}${params}`
    )

    return response.data
  },
}
